nucleus/dataset.py

```diff
@@ -1414,11 +1429,76 @@ def items_and_annotation_generator(
             endpoint=f"dataset/{self.id}/exportForTrainingPage",
             result_key=EXPORT_FOR_TRAINING_KEY,
             page_size=10000,  # max ES page size
+            query=query,
         )
         for data in json_generator:
             for ia in convert_export_payload([data], has_predictions=False):
                 yield ia
 
+    def items_and_annotation_chip_generator(
+        self,
+        chip_size: int,
+        stride_size: int,
+        cache_directory: str,
+        query: Optional[str] = None,
+    ) -> Iterable[Dict[str, str]]:
+        """Provides a generator of chips for all DatasetItems and BoxAnnotations in the dataset.
+
+        Args:
+            chip_size: The size of the image chip
+            stride_size: The distance to move when creating the next image chip.
+              When stride is equal to chip size, there will be no overlap.
+              When stride is equal to half the chip size, there will be 50 percent overlap.
+            cache_directory: The s3 or local directory to store the image and annotations of a chip.
+              s3 directories must be in the format s3://s3-bucket/s3-key
+            query: Structured query compatible with the `Nucleus query language <https://nucleus.scale.com/docs/query-language-reference>`_.
+
+        Returns:
+            Generator where each element is a dict containing the location of the image chip (jpeg) and its annotations (json).
+            ::
+
+                Iterable[{
+                    "image_location": str,
+                    "annotation_location": str
+                }]
+        """
+        json_generator = paginate_generator(
+            client=self._client,
+            endpoint=f"dataset/{self.id}/exportForTrainingPage",
+            result_key=EXPORT_FOR_TRAINING_KEY,
+            page_size=10000,  # max ES page size
+            query=query,
+            chip=True,
+        )
+        for item in json_generator:
+            image = fetch_image(item[ITEM_KEY][PROCESSED_URL_KEY])
+            w, h = image.size
+            annotations = item[BOX_TYPE]
+            item_ref_id = item[ITEM_KEY][REFERENCE_ID_KEY]
+            for x0, y0 in generate_offsets(w, h, chip_size, stride_size):
+                x0, y0 = int(x0), int(y0)
+                ref_id = f"{cache_directory}/{chip_size}/{stride_size}/{item_ref_id}_{x0}_{y0}"
```

Comment:
I think this can probably just be `f"{cache_directory}/{item_ref_id}_{x0}_{y0}`. In a world in which we we have one dataset of chip_size=1024 and stride_size=512 and then another dataset of chip_size=512 and stride_size=512, there will be duplicate chips so we'd want to be able utilize the available cache from one to the other if possible

Reply:
in that case I think it should be `f"{cache_directory}/{item_ref_id}_{x0}_{y0}_{x1}_{y1}` otherwise the top left of the image would be the same in both cases which is not correct